File: python/queue.py
#!/usr/bin/python

"""
    A queue class.
    Author: Daniel Ribeirinha-Braga
"""


class Queue:
    def __init__(self):
        self.queue = []

    def enqueue(self, item):
        self.queue.append(item)

    def dequeue(self):
        if self.is_empty():
            raise Exception("Queue is empty")
        return self.queue.pop(0)

    def peek(self):
        if self.is_empty():
            raise Exception("Queue is empty")
        return self.queue[0]

    def is_empty(self):
        return len(self.queue) == 0

    def size(self):
        return len(self.queue)

File: java/Queue.java
import java.util.LinkedList;

public class Queue<T> {
  private LinkedList<T> queue;

  public Queue() {
      queue = new LinkedList<>();
  }

  public void enqueue(T item) {
      queue.addLast(item);
  }

  public T dequeue() {
      if (isEmpty()) {
          throw new IllegalStateException("Queue is empty");
      }
      return queue.removeFirst();
  }

  public T peek() {
      if (isEmpty()) {
          throw new IllegalStateException("Queue is empty");
      }
      return queue.getFirst();
  }

  public int size() {
      return queue.size();
  }

  public boolean isEmpty() {
      return queue.isEmpty();
  }
}

